src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp

```diff
@@ -29,57 +26,50 @@ void Tactic::updateRobot(const Robot &robot)
 
 std::unique_ptr<Intent> Tactic::getNextIntent()
 {
+    std::unique_ptr<Intent> next_intent = nullptr;
     if (!robot)
     {
         LOG(WARNING) << "Requesting the next Intent for a Tactic without a Robot assigned"
                      << std::endl;
-        return std::unique_ptr<Intent>{};
     }
-
-    auto next_intent = getNextIntentHelper();
-    if (done_ && loop_forever)
+    else
     {
-        // If the tactic is done and is supposed to loop forever, we re-create the
-        // intent_sequence which "restarts" the coroutine. We then run the coroutine
-        // again, and return the result from the restarted coroutine rather than the
-        // old one. This way, any callers of this function won't accidentally get a
-        // nullptr returned for a single call (which could come from the "old" coroutine)
-        // when this Tactic restarts
-        intent_sequence.reset();
-        intent_sequence = std::make_unique<intent_coroutine::pull_type>(
-            [this](intent_coroutine::push_type &yield) {
-                return this->calculateNextIntentWrapper(yield);
-            });
         next_intent = getNextIntentHelper();
+        if (done_ && loop_forever)
+        {
+            // Re-start the intent sequence by re-creating it
+            intent_sequence = IntentCoroutine::pull_type(
+                boost::bind(&Tactic::calculateNextIntentWrapper, this, _1));
+            next_intent = getNextIntentHelper();
+        }
```

Comment:
```
        next_intent = getNextIntentHelper();
        if (done_ && loop_forever)
        {
            // Re-start the intent sequence by re-creating it
            intent_sequence = IntentCoroutine::pull_type(
                boost::bind(&Tactic::calculateNextIntentWrapper, this, _1));
            next_intent = getNextIntentHelper();
        }
```
can be simplified to:
```
        if (done_ && loop_forever)
        {
            // Re-start the intent sequence by re-creating it
            intent_sequence = IntentCoroutine::pull_type(
                boost::bind(&Tactic::calculateNextIntentWrapper, this, _1));
        }
        next_intent = getNextIntentHelper();
```

Reply:
Not quite. It's this way because the `getNextIntentHelper` function sets the `done` flag. If the first call results in the tactic being done, that also means a nullptr would be returned. We want to avoid returning a nullptr at any point while we are looping forever, so we check for this and immediately restart (in the current code).

The change you suggested would allows a single nullptr to be returned in between tactic restarts, because the `next_intent` would be returned as a nullptr and not caught until the next call.

I've added comments to clarify this.